File: lib/java/codedifferently-instructional/instructional-lib/src/main/java/com/codedifferently/instructional/quiz/QuizQuestion.java
package com.codedifferently.instructional.quiz;

import java.util.Objects;
import java.util.Scanner;

public class QuizQuestion {

  private int questionNumber;

  private String question;

  private String correctAnswer;

  public QuizQuestion(int questionNumber, String question, String correctAnswer) {
    this.questionNumber = questionNumber;
    this.question = question;
    this.correctAnswer = correctAnswer;
  }

  public int getQuestionNumber() {
    return questionNumber;
  }

  public String getQuestionPrompt() {
    return question;
  }

  public String getAnswer() {
    return Objects.toString(correctAnswer, "");
  }

  public void setAnswer(String answer) {
    this.correctAnswer = answer;
  }

  public void promptForAnswer(Scanner scanner) {
    promptForAnswer(scanner, false);
  }

  public void promptForAnswer(Scanner scanner, boolean requireAnswer) {
    String response = null;

    do {
      // If a response was given but not accepted, announce that.
      if (response != null) {
        System.out.println("Value not accepted, try again.\n");
      }

      // Get a response.
      String prompt =
          ">> Your answer " + (requireAnswer ? "(required)" : "(or 's' to skip)") + ": ";
      System.out.print(prompt);
      response = scanner.next().trim();

      // If the user wants to skip, let them.
      if (!requireAnswer && response.toUpperCase().equals("S")) {
        this.setAnswer(null);
        return;
      }
    } while (response.equals(""));

    // Set the question answer.
    this.setAnswer(response);
  }
}

File: lib/typescript/codedifferently-instructional/src/quiz_question.ts
export class QuizQuestion {
  private questionNumber: number;
  private question: string;
  private correctAnswer: string | null;

  constructor(questionNumber: number, question: string, correctAnswer: string) {
    this.questionNumber = questionNumber;
    this.question = question;
    this.correctAnswer = correctAnswer;
  }

  public getQuestionNumber(): number {
    return this.questionNumber;
  }

  public getQuestionPrompt(): string {
    return this.question;
  }

  public getAnswer(): string {
    return this.correctAnswer ?? '';
  }

  public setAnswer(answer: string | null): void {
    this.correctAnswer = answer;
  }

  public promptForAnswer(scanner: Scanner, requireAnswer = false): void {
    let response: string | null = null;

    do {
      // If a response was given but not accepted, announce that.
      if (response !== null) {
        console.log("Value not accepted, try again.\n");
      }

      // Get a response.
      const prompt = `>> Your answer ${requireAnswer ? "(required)" : "(or 's' to skip)"}: `;
      process.stdout.write(prompt);
      response = scanner.next().trim();

      // If the user wants to skip, let them.
      if (!requireAnswer && response.toUpperCase() === "S") {
        this.setAnswer(null);
        return;
      }
    } while (requireAnswer && (response === '' || response.toUpperCase() === 'S'));

    // Set the question answer.
    this.setAnswer(response);
  }
}

// Scanner class to simulate Java's Scanner for user input
class Scanner {
  public next(): string {
    return require('readline-sync').question('');
  }
}
